### PR TITLE
No issue: Offline Audio Buffer rendering.

### PR DESCRIPTION
**Pull request by pablocubico**
Hi!

I needed to integrate ToneJS to be able to code some audio stuff into videos, without having to provide any pre-made audio files.

Some changes are probably unwanted, but I think it's a good starting point for a very powerful feature, which could allow users to code sound events using Remotion. In my case, I'm working on some music theory videos.

This PR of course adds `ToneJS` and `audiobuffer-to-wav` as dependencies, but they aren't needed, it was just easy to tweak the main example directly. 😅 

The overall idea is to be able to render an AudioBuffer as a file so it can go into Remotion's assets pipeline as any other file. So it just introduces some very small (but very opinionated!) changes. 

This is the output it produces: https://drive.google.com/file/d/1LXWcQqziPUiC0LeuRSYISiLC9yPbIkV-/view?usp=sharing

That's it, I think it's pretty self-explanatory, I'll leave some comments in the code. 

Please share your thoughts!

### CODE REVIEW

File: packages/example/src/ToneJSTesting/ToneJSExample.tsx
Comment (pablocubico) [inline packages/example/src/ToneJSTesting/ToneJSExample.tsx]: This was just a sample I found, it probably needs to be extracted and maybe ensure this is the fastest method.

File: packages/example/src/ToneJSTesting/ToneJSExample.tsx
Comment (pablocubico) [inline packages/example/src/ToneJSTesting/ToneJSExample.tsx]: I mean, if `<Audio />` supports an AudioBuffer as source, it should decide whether to play it or render it offline depending on the environment. Maybe the Audio component itself could know about encoding an AudioBuffer to base64 🤔

Comment (pablocubico) [reply packages/example/src/ToneJSTesting/ToneJSExample.tsx]: I'm probably doing some extra steps here, converting the `AudioBuffer` to `wav` into a `blob:` seems more straightforward. I'll dig a bit more.

**Comment by JonnyBurger** [comment]
Thanks for the PR! This is very cool 😃

- Happy to add support for data URL assets!
- example/package-lock.json and the .d.ts artifacts should be removed
- Agree with that `wav` is a bit too hardcoded. It seems like we can do it for all base 64 data?

PS I meant to comment and accidentially closed the PR, of course meant not to do that :)

**Comment by pablocubico** [comment]
Great to hear! 🙌 

 I'll clean this PR to make it more agnostic, remove deps & artifacts, and iterate again.
 
 Stay tuned!

**Comment by pablocubico** [comment]
Updated with some cleanups: 
* Removed unwanted artifacts.
* Removed ToneJS as a dependency. 
* Made the `data:` portion of `downloadAsset` a bit more agnostic. 

Unexpected side-effects:
* After removing ToneJS, the note played sounds wobbly in the video file, as if two frequencies were sounding together. Currently investigating.

**Comment by pablocubico** [comment]
Update + note to self: the audio file generated looks & sounds good in the temp directory, it just sounds wobbly on the video. Checked both ToneJS & Vanilla WAV outputs with Audacity & MediaInfo, this is the MediaInfo output for both of them:

ToneJS version:
```
General
Complete name                            : with-tone.wav
Format                                   : Wave
File size                                : 861 KiB
Duration                                 : 5 s 0 ms
Overall bit rate mode                    : Constant
Overall bit rate                         : 1 411 kb/s

Audio
Format                                   : PCM
Format settings                          : Little / Signed
Codec ID                                 : 1
Duration                                 : 5 s 0 ms
Bit rate mode                            : Constant
Bit rate                                 : 1 411.2 kb/s
Channel(s)                               : 2 channels
Sampling rate                            : 44.1 kHz
Bit depth                                : 16 bits
Stream size                              : 861 KiB (100%)
```

Vanilla version:
```
General
Complete name                            : with-vanilla.wav
Format                                   : Wave
File size                                : 345 KiB
Duration                                 : 2 s 0 ms
Overall bit rate mode                    : Constant
Overall bit rate                         : 1 411 kb/s

Audio
Format                                   : PCM
Format settings                          : Little / Signed
Codec ID                                 : 1
Duration                                 : 2 s 0 ms
Bit rate mode                            : Constant
Bit rate                                 : 1 411.2 kb/s
Channel(s)                               : 2 channels
Sampling rate                            : 44.1 kHz
Bit depth                                : 16 bits
Stream size                              : 345 KiB (100%)
```

Still debugging 👀 ☕

**Comment by pablocubico** [comment]
Wobbly audio fixed! ✅ 

Apparently the default value for the GainNode's gain (which is `1`) caused the mix to be wobbly. ToneJS applies an envelope to the Synth so the waveform was not as aggressive as the pure oscillator vanilla example. Using `.5` for the gainNode produced clear output. Maybe it's something to watch out and log as an issue for the future (?).

Known issues & things to consider from the top of my head:
* The generated audio does not show in the preview's timeline.
* Consider implications of the `<Audio />` API (startFrom, endAt).
* Consider implications of other timeline design features (like Sequence). 
* Better way to determine audio length?
* Ways to sync audio events with animation? (maybe in a separate ToneJS example repo, syncing BPMs/FPS, etc)
* Any security issues for the `data:` file dump? Maybe allowing certaing mimeTypes? 
* `audiobuffer-to-wav` is [very simple](https://github.com/Jam3/audiobuffer-to-wav/blob/master/index.js), maybe it's worth creating a helper and remove the dependency... or not, looks healthy in NPM anyway (7K downloads/week).

File: packages/example/package.json
Comment (pablocubico) [inline packages/example/package.json]: No longer needed here, will remove.

**Comment by pablocubico** [comment]
FYI: I created a separate branch with a ToneJS example with synchronized audio & video (which was my main goal), it's here: https://github.com/pablocubico/remotion/commit/b34a602da01e823ada86b9bbc4e87b9c8f43b45e

The output is: https://drive.google.com/file/d/1jWNjxpyKd8gOlg7TnjFcGqLTsGRGD4-p/view?usp=sharing

💥 🎵

**Comment by JonnyBurger** [comment]
@pablocubico! Oh cool! I was bummed the note.js example went away 😊 

I am refactoring it a bit and moving the audio buffer to base64 conversion out of the core and into the @remotion/media-utils  package. After that I am willing to merge it! 😃

**Comment by pablocubico** [comment]
@JonnyBurger super, I was just looking at your changes, looks amazing! 🥇

**Comment by JonnyBurger** [comment]
I'm done with the refactoring and if you give me a 👍  I will merge it and release it today!

I'm trying to recreate Jingle Bells in Tone.JS now 😁 🎄

**Comment by pablocubico** [comment]
@JonnyBurger Yes sir, here you go: 👍👍👍  

I've been reading your commits, looks great to me! Also, the approach is quite simple, isn't it? All high-level stuff.

I seriously needed this, best christmas gift possible!! 🎄 🎅

**Comment by JonnyBurger** [comment]
You are right! It's fantastic.

Thanks a lot for working on this. I am merging it now once CI passes.
I also did a Jingle Bells Demo in Remotion, gonna post it tomorrow! Do you have Twitter and Instagram handles where I can tag you?

**Review by JonnyBurger — APPROVED** [review]

**Comment by pablocubico** [comment]
Sure, both: 
https://twitter.com/pablocubico
[instagram.com/pablocubico/](https://www.instagram.com/pablocubico/)

Thanks so much!

**Pull request merged**
